File: static/js/parser/Ranges.js

class Range {
    constructor() {
        if(this.constructor == Range) {
            throw new Error("Abstract class cannot be instantiated.");
        }
    }

    getValue() {
        throw new Error("Abstract method has not been implemented.");
    }
}

class RealRange extends Range {
    constructor(min, max, minInclusive, maxInclusive, dataRange) {
        super();
        this.min = min;
        this.max = max;
        this.minInclusive = minInclusive;
        this.maxInclusive = maxInclusive;
        this.dataRange = dataRange;
    }

    getValue() {
        if(this.dataRange != null) {
            let index = Math.round(Math.random() * (this.dataRange.length - 1));
        
            return this.dataRange[index];
        }

        let pick = Math.random() * (this.max - this.min) + this.min;

        if(!this.minInclusive) {
            while(pick == this.min) {
                pick = Math.random() * (this.max - this.min) + this.min;
            }
        }

        if(!this.maxInclusive) {
            while(pick == this.max) {
                pick = Math.random() * (this.max - this.min) + this.min;
            }
        }

        return pick;
    }
}

class IntegerRange extends Range {
    constructor(min, max, minInclusive, maxInclusive, dataRange) {
        super();

        this.min = min;
        this.max = max;
        this.minInclusive = minInclusive;
        this.maxInclusive = maxInclusive;
        this.dataRange = dataRange;
    }

    getValue() {
        if(this.dataRange != null) {
            let index = Math.round(Math.random() * (this.dataRange.length - 1));
        
            console.log(index);

            return this.dataRange[index];
        }

        var min = this.min;
        var max = this.max;

        if(!this.minInclusive) {
            min += 1;
        }

        if(!this.maxInclusive) {
            max -= 1;
        }

        let pick = Math.round(Math.random() * (max - min) + min);

        return pick;
    }
}

class CharacterRange extends Range {
    constructor(min, max, minInclusive, maxInclusive, dataRange) {
        super();
        this.min = min.charCodeAt(0);
        this.max = max.charCodeAt(0);
        this.minInclusive = minInclusive;
        this.maxInclusive = maxInclusive;
        this.dataRange = dataRange;
    }

    getValue() {
        if(this.dataRange != null) {
            let index = Math.round(Math.random() * (this.dataRange.length - 1));
        
            return this.dataRange[index];
        }
        
        var min = this.min;
        var max = this.max;

        if(!this.minInclusive) {
           min += 1;
        }

        if(!this.maxInclusive) {
            max -= 1;
        }

        let pick = Math.round(Math.random() * (max - min) + min);

        return String.fromCharCode(pick);
    }
} 

class StringRange extends Range {
    constructor(min, max, minInclusive, maxInclusive, dataRange) {
        super();
        this.min = min;
        this.max = max;
        this.minInclusive = minInclusive;
        this.maxInclusive = maxInclusive;
        this.dataRange = dataRange;
    }

    getValue() {
        if(this.dataRange != null) {
            let index = Math.round(Math.random() * (this.dataRange.length - 1));
        
            return this.dataRange[index];
        }

        var min = this.min;
        var max = this.max;

        if(!this.minInclusive) {
            min += 1;
        }

        if(!this.maxInclusive) {
            max -= 1;
        }

        let pick = Math.round(Math.random() * (max - min) + min);
        let result = "";
        var characters = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789';

        for (let i = 0; i < pick; i++) {
            result += characters.charAt(Math.floor(Math.random() * characters.length));
        }

        return result;
    }
} 

File: Main/testGeneration/ranges.py
from random import uniform
from random import randint

class Range:

    def getValue(self):
        """Bae method for concrete ranges

        Returns:
            _type_: A random generated value
        """

class RealRange(Range):

    def __init__(self, min, max, minInclusive, maxInclusive, dataRange):
        self.__min = min
        self.__max = max
        self.__minInclusive = minInclusive
        self.__maxInclusive = maxInclusive
        self.__dataRange = dataRange

    def getValue(self):
        if self.__dataRange:
            pick = round(uniform(0, len(self.__dataRange)))

            return self.__dataRange[pick]
    
        pick = uniform(self.__min, self.__max)

        if not self.__minInclusive:
            while pick == self.__min:
                pick = uniform(self.__min, self.__max)
        elif not self.__maxInclusive:
            while pick == self.__max:
                pick = uniform(self.__min, self.__max)
        
        return pick

class IntegerRange(Range):
    def __init__(self, min, max, minInclusive, maxInclusive, dataRange):
        if not minInclusive:
            min += 1
        if maxInclusive:
            max += 1

        if not dataRange:
            self.__range = range(min, max)
        else:
            self.__range = dataRange

    def getValue(self):
        pick = randint(0, len(self.__range)-1)

        return self.__range[pick]

class CharacterRange(Range):
    def __init__(self, min, max, minInclusive, maxInclusive, dataRange):
        if not minInclusive:
            min = chr(ord(min) + 1) 
        if maxInclusive:
            max = chr(ord(max) + 1)

        self.__range = dataRange

        if not dataRange:
            self.__range = []
            for i in range(0, ord(max) - ord(min)):
                self.__range.append(chr(ord(min) + i))

    def getValue(self):
        pick = randint(0, len(self.__range)-1)

        return self.__range[pick]

class StringRange(Range):
    def __init__(self, min, max, minInclusive, maxInclusive, dataRange):
        if not minInclusive:
            min += 1
        if maxInclusive:
            max += 1

        if not dataRange:
            self.__rangeType = "Integer"
            self.__range = range(min, max)
        else:
            self.__rangeType = "String"
            self.__range = dataRange

    def getValue(self):
        pick = randint(0, len(self.__range)-1)

        if self.__rangeType == "String":
            return self.__range[pick]

        random_string = ""
        for i in range(self.__range[pick]):
            flip_bit = randint(0, 1)
            random_string += chr(randint(97, 122) - 32 * flip_bit)

        return random_string